### PR TITLE
feat: added s3 artifact

### LINKED ISSUE #37: Add S3ArtifactService for Amazon S3 artifact storage support
**Is your feature request related to a problem? Please describe.**

ADK currently provides `GcsArtifactService` for Google Cloud Storage and `FileArtifactService` for local development, but there is no native artifact service for Amazon S3. This creates friction for developers and organizations using AWS infrastructure, as they must either:
- Use `FileArtifactService` (not production-ready, limited scalability)
- Use `GcsArtifactService` (requires cross-cloud setup, additional complexity)
- Implement their own S3 service from scratch (time-consuming, error-prone)

For AWS-native deployments (EC2, ECS, Lambda, EKS), developers need a first-class S3 artifact storage solution that integrates seamlessly with their existing infrastructure and IAM policies.

**Describe the solution you'd like**

Implement `S3ArtifactService` as a production-ready artifact storage backend with the following characteristics:

**Core Features:**
- Extends `BaseArtifactService` following the same patterns as `GcsArtifactService`
- Supports session-scoped artifacts: `{app}/{user}/{session}/{filename}/{version}`
- Supports user-scoped artifacts: `{app}/{user}/user/{filename}/{version}` (using `user:` prefix)
- Automatic version management (0, 1, 2, ...)
- Custom metadata support via S3 object metadata
- URL encoding for special characters in filenames
- Async/sync pattern using `asyncio.to_thread`

**Technical Implementation:**
- Uses boto3 for S3 interactions
- Added as optional dependency in `pyproject.toml` under `extensions`
- No breaking changes to existing code
- Full feature parity with `GcsArtifactService`

**Developer Experience:**
```python
from google.adk.artifacts import S3ArtifactService

artifact_service = S3ArtifactService(
    bucket_name="my-adk-artifacts",
    region_name="us-east-1"
)

runner = Runner(agent=my_agent, artifact_service=artifact_service)
```

**Describe alternatives you've considered**

1. **Using GcsArtifactService with cross-cloud setup**
   - Requires managing credentials across cloud providers
   - Additional latency for cross-cloud data transfer
   - Increased costs for egress between clouds
   - More complex IAM/permission management

2. **Using FileArtifactService for production**
   - Not designed for production (single machine, no replication)
   - Limited by disk space
   - No durability guarantees (99.999999999% for S3 vs. disk failure)
   - Requires persistent volumes in containerized environments

3. **Implementing custom S3 wrapper**
   - Every team duplicates effort
   - No standardization across projects
   - Potential bugs and inconsistencies
   - Maintenance burden on each team

4. **Using external artifact storage services**
   - Additional service dependency
   - Extra cost
   - Learning curve for another tool
   - Less control over infrastructure

### PR DESCRIPTION
**Pull request by miyannishar**
## Link to Issue or Description of Change

**1. Link to an existing issue (if applicable):**

- Closes: #37 

**Problem:**

ADK main repo currently provides `GcsArtifactService` for Google Cloud Storage and `FileArtifactService` for local development, but there is no native artifact service for Amazon S3. This creates friction for developers and organizations using AWS infrastructure, as they must either:
- Use `FileArtifactService` (not production-ready, limited scalability)
- Use `GcsArtifactService` (requires cross-cloud setup, additional complexity)
- Implement their own S3 service from scratch (time-consuming, error-prone)

For AWS-native deployments (EC2, ECS, Lambda, EKS), developers need a first-class S3 artifact storage solution that integrates seamlessly with their existing infrastructure and IAM policies.

**Solution:**

Implemented `S3ArtifactService` as a production-ready artifact storage backend that:

- Extends `BaseArtifactService` following the same patterns as `GcsArtifactService`
- Supports session-scoped artifacts: `{app}/{user}/{session}/{filename}/{version}`
- Supports user-scoped artifacts: `{app}/{user}/user/{filename}/{version}` (using `user:` prefix)
- Automatic version management (0, 1, 2, ...)
- Custom metadata support via S3 object metadata
- URL encoding for special characters in filenames
- Async/sync pattern using `asyncio.to_thread`
- Works with S3-compatible services (MinIO, DigitalOcean Spaces, etc.)
- Added `s3://` URI scheme support in service registry
- boto3 added as optional dependency in `extensions`

**Key Files Changed:**
- `src/google/adk/artifacts/s3_artifact_service.py` (NEW - ~612 lines)
- `src/google/adk/artifacts/__init__.py` (export S3ArtifactService)
- `src/google/adk/cli/service_registry.py` (register s3:// URI scheme)
- `pyproject.toml` (add boto3>=1.28.0 to extensions)
- `tests/unittests/artifacts/test_artifact_service.py` (add S3 tests)

### Unit Tests

- [x] I have added or updated unit tests for my change.
- [x] All unit tests pass locally. _(Run tests using guide below)_

**Unit Test Summary:**

Added comprehensive unit tests for S3ArtifactService:

1. **Mock Infrastructure:**
   - `MockS3Client`: Simulates boto3 S3 client
   - `MockS3Bucket`: Simulates S3 bucket
   - `MockS3Object`: Simulates S3 objects with metadata
   - `mock_s3_artifact_service()`: Factory function for tests

2. **Test Coverage:**
   - ✅ `test_load_empty[S3]` - Loading non-existent artifacts
   - ✅ `test_save_load_delete[S3]` - Basic CRUD operations
   - ✅ `test_list_keys[S3]` - Listing artifact keys
   - ✅ `test_list_versions[S3]` - Version management
   - ✅ `test_list_keys_preserves_user_prefix[S3]` - User-scoped artifacts
   - ✅ `test_list_artifact_versions_and_get_artifact_version[S3]` - Metadata operations
   - ✅ `test_list_artifact_versions_with_user_prefix[S3]` - User-scoped metadata
   - ✅ `test_get_artifact_version_artifact_does_not_exist[S3]` - Error handling
   - ✅ `test_get_artifact_version_out_of_index[S3]` - Edge cases

3. **Test Execution:**
   ```bash
   # Run S3-specific tests
     pytest tests/unittests/artifacts/test_s3_artifact_service.py -v
   
   # Expected: 9 tests pass
   ```

4. **Test Results:** _(Fill this in after running tests)_

image

### Manual End-to-End (E2E) Tests

**Setup:**
```python
#!/usr/bin/env python3
"""Real S3 integration test - uploads files, waits, then deletes one."""
import asyncio
import os
import time

try:
    from dotenv import load_dotenv
    load_dotenv()
except ImportError:
    pass

from google.adk_community.artifacts import S3ArtifactService
from google.genai import types


async def main():
    bucket = os.getenv("TEST_S3_BUCKET")
    if not bucket:
        print("❌ Set TEST_S3_BUCKET in .env file")
        return
    
    region = os.getenv("AWS_REGION", "us-east-1")
    print(f"S3 Integration Test: {bucket} ({region})\n")
    
    # Initialize service
    service = S3ArtifactService(bucket_name=bucket, region_name=region)
    
    # Upload two files
    print("📤 Uploading two files...")
    file1 = types.Part(text="Report data for Q4 2024")
    file2 = types.Part(text="Meeting notes from Dec 10")
    
    v1 = await service.save_artifact(
        app_name="test-app", user_id="user1", session_id="session1",
        filename="report.txt", artifact=file1,
        custom_metadata={"type": "report", "quarter": "Q4"}
    )
    print(f"  ✅ Uploaded report.txt (version {v1})")
    
    v2 = await service.save_artifact(
        app_name="test-app", user_id="user1", session_id="session1",
        filename="notes.txt", artifact=file2,
        custom_metadata={"type": "notes", "date": "2024-12-10"}
    )
    print(f"  ✅ Uploaded notes.txt (version {v2})")
    
    # List files
    files = await service.list_artifact_keys(
        app_name="test-app", user_id="user1", session_id="session1"
    )
    print(f"\n📋 Files in S3: {files}")
    
    # Wait briefly
    print(f"\n⏳ Waiting 60 seconds...")
    time.sleep(60)
    
    # Verify both files exist
    print("\n🔍 Verifying files...")
    loaded1 = await service.load_artifact(
        app_name="test-app", user_id="user1", session_id="session1",
        filename="report.txt"
    )
    loaded2 = await service.load_artifact(
        app_name="test-app", user_id="user1", session_id="session1",
        filename="notes.txt"
    )
    print(f"  ✅ report.txt: {loaded1.inline_data.data.decode()[:30]}...")
    print(f"  ✅ notes.txt: {loaded2.inline_data.data.decode()[:30]}...")
    
    # Delete one file
    print("\n🗑️  Deleting report.txt...")
    await service.delete_artifact(
        app_name="test-app", user_id="user1", session_id="session1",
        filename="report.txt"
    )
    
    # List files again
    files_after = await service.list_artifact_keys(
        app_name="test-app", user_id="user1", session_id="session1"
    )
    print(f"📋 Files after deletion: {files_after}")
    
    # Cleanup - delete remaining file
    print("\n🧹 Cleanup: deleting notes.txt...")
    await service.delete_artifact(
        app_name="test-app", user_id="user1", session_id="session1",
        filename="notes.txt"
    )
    
    print("\n✅ Integration test completed successfully!")


if __name__ == "__main__":
    asyncio.run(main())
```

Loom Video of testing: https://www.loom.com/share/8617ff91cdb14dd98edb5176f8f3cec2

## Checklist

- [x] I have performed a self-review of my own code.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have added tests that prove my fix is effective or that my feature works.
- [x] Unit tests pass locally with my changes. _(Run tests - see guide)_
- [x] I have manually tested my changes end-to-end. _(Optional but recommended)_
- [x] Any dependent changes have been merged and published in downstream modules. _(N/A - new feature)_

### Design Decisions

1. **Optional Dependency:** boto3 is in `extensions` to avoid breaking existing installations
2. **URL Encoding:** Handles special characters in filenames (e.g., `/`, `:`, spaces)
3. **Async Pattern:** Uses `asyncio.to_thread` matching `GcsArtifactService` pattern
4. **Service Registry:** Follows same pattern as `gs://` for consistency
6. **Feature Parity:** All methods from `BaseArtifactService` implemented

### Comparison with Existing Services

| Feature | S3ArtifactService | GcsArtifactService | FileArtifactService |
|---------|-------------------|---------------------|---------------------|
| Cloud Provider | AWS | GCP | Local |
| Durability | 99.999999999% | 99.999999999% | Disk-dependent |
| Scalability | Unlimited | Unlimited | Disk-limited |
| Production Ready | ✅ Yes | ✅ Yes | ⚠️ Dev/Test Only |
| URL Encoding | ✅ Yes | ❌ No | ❌ No |

### CODE REVIEW

File: src/google/adk_community/artifacts/s3_artifact_service.py [hunk 1]
```diff
@@ -0,0 +1,694 @@
...
+      if not versions:
+        return None
+      version = max(versions)
+
+    object_key = self._get_object_key(
+        app_name, user_id, filename, version, session_id
+    )
+
+    try:
+      response = self.s3_client.get_object(
```
Comment (DeanChensj) [inline src/google/adk_community/artifacts/s3_artifact_service.py]: Is there any async version of method in s3_client?  If so, can we use them directly so we don't need to use asyncio?

**Comment by sasha-gitg** [comment]
Closing this for lack of response from the author.

**Comment by miyannishar** [comment]
Hey! I've opened a fresh PR at #115 that supersedes this one.

The new PR combines the best of this PR and #29:
- **Native async** via `aioboto3` (addresses the `asyncio.to_thread` concern raised in review)
- **Atomic versioning** via `IfNoneMatch` conditional writes (solves the race condition)
- **Optional `[s3]` dependency** (not a hard dep on `aioboto3`)
- **No sample agent** (removed as requested by @DeanChensj)
- **Cleaned up README** (contributing section removed)
- **Comprehensive test suite** (13 tests with full async mock infrastructure)

Thanks for the feedback on this PR — it directly informed the improvements in the new one!
